### DISCUSSION TITLE
PackageRules order

### DISCUSSION THREAD

**Discussion by hieroshima** [renovatebot/renovate#8487] [Request Help] [answered]
Hi,

I have a default config stored in the repository from where all other renovate configuration extends. This is the default config:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    ":ignoreModulesAndTests",
    ":semanticCommits",
    ":disableRateLimiting",
    ":automergeDisabled",
    ":maintainLockFilesDisabled",
    ":ignoreUnstable",
    ":semanticCommitTypeAll(MyBot)",
    ":preserveSemverRanges"
  ],
  "separateMultipleMajor": false,
  "separateMinorPatch": false,
  "packageRules": [
    {
      "depTypeList": ["devDependencies"],
      "branchPrefix": "devDependencies/all-",
      "groupName": "all",
      "groupSlug": "devDependencies",
      "labels": ["devDependency"]
    },
    {
      "depTypeList": ["dependencies"],
      "branchPrefix": "dependencies/update-{{depName}}-",
      "groupName": "dependencies",
      "groupSlug": "dependencies",
      "labels": ["dependency"]
    },
    {
      "updateTypes": ["major"],
      "addLabels": ["major"]
    },
    {
      "updateTypes": ["minor"],
      "addLabels": ["minor"]
    },
    {
      "updateTypes": ["patch"],
      "addLabels": ["patch"]
    }
  ],
  "commitBodyTable": true,
  "prHeader": "{{#if isMajor}}# {{depName}} v{{{newMajor}}} &ndash;&ndash; :bangbang: MAJOR UPDATE! :bangbang:{{/if}}",
  "commitMessageSuffix": "{{#if isMajor}}[MAJOR]{{/if}}"
}

```

All repositories extend from that one base file. I have now repository with angular and ngrx where I would like to have all (minor,major,patch) updates in one MR (one for angular and one for ngrx) - rest libraries should follow the rules from base file. I prepared that configuration in this repo:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["gitlab>renovate-config:app-renovate-default"],
  "packageRules": [
    {
    "depTypeList": ["devDependencies", "dependencies"]
    "packagePatterns": ["^angular"],
    "groupName": "angular",
    "groupSlug": "angular",
    "branchPrefix": "dependencies/angular-update-",
    "addLabels": ["angular"]
    },
    {
    "depTypeList": ["devDependencies", "dependencies"]
    "packagePatterns": ["^ngrx"],
    "groupName": "ngrx",
    "groupSlug": "ngrx",
    "branchPrefix": "dependencies/ngrx-update-",
    "addLabels": ["ngrx"]
    }
  ]
}
```

The problem is that angular have devDependencies and dependencies and both types are matched also in the base rules from the base file. What is the simplest way to achieve what I want? Should I exclude angular and ngrx from the packageRules on the base file?

Thanks for any kind of help :)

**Comment by rarkins** [comment, marked as answer]
When you say:

> all (minor,major,patch) updates in one MR (one for angular and one for ngrx)

I don't know for sure what you mean.

Do you mean you no longer want to separate major vs minor vs patch for these dependencies? e.g. if there's a major update available then don't raise any separate minor or patch PR? Or do you still want separate PRs for minor/patch/major?

Do not change `branchPrefix` in `packageRules` - it is reserved for the top level. It is `groupSlug` which determines groupings. as it is added to the `branchPrefix`.

**Reply by hieroshima** [reply]
I would like to have updates for angular in one PR without separating for minor/patch/major and the same for ngrx. For rest dependencies, I would like to have rules from the base file which group all devDeps, but separate not dev dependancies.

**Reply by rarkins** [reply]
So this sounds right?
- For angular and ngrx packages, no longer separate major from non-major updates. At most one update per package, even if it's a major. In other words: ignore minor updates if a major is available
- Have a single "Update angular packages" PR and a single "Update ngrx packages" PR which combine all updates in a single PR (could be mix or minor and major for different packages)

**Reply by hieroshima** [reply]
That is exactly right :)

**Reply by rarkins** [reply]
You already have the first point satisfied in your base config:

```
  "separateMultipleMajor": false,
  "separateMinorPatch": false,
```

To achieve the second part, you must remove them from the existing grouping, similar to how you had it:

```
  "packageRules": [
    {
    "packagePatterns": ["^angular"],
    "groupName": "angular",
    "groupSlug": "angular",
    "addLabels": ["angular"]
    },
    {
    "packagePatterns": ["^ngrx"],
    "groupName": "ngrx",
    "groupSlug": "ngrx",
    "addLabels": ["ngrx"]
    }
  ]
```

I think the above should work, but if not then I'd need you to reproduce the problem in a public repo so that I can debug it.

Please also remove the `branchPrefix` from your base config packageRules.

**Reply by hieroshima** [reply]
I've did that and removed branchPrefix, but the problem is that angular has something in devDependancies and it is included in the first PR where all devDependancies are mixed, so in fact I have now 4 PR: all major devDep, all minor/patch devDep, all major dep and all minor/patch dep. I don't have separate for angular/ngrx

**Reply by rarkins** [reply]
Which version of Renovate are you running?

**Reply by hieroshima** [reply]
24.27.1

**Reply by rarkins** [reply]
The packageRules in your repo should be sorted after the packageRules in your base preset, and therefore override them. Please run Renovate with the option `--print-config=true` so that it dumps your full config and you can confirm the rule ordering.

Also, please take note of the branch names in the PRs you are seeing so that we can confirm that they are as expected.

**Reply by hieroshima** [reply]
It does not print any config when I tried `renovate --print-config=true` or `renovate --print-config`

Branch names are also not good I think:

- branch=renovate/all-dependencies
- branch=renovate/all-devdependencies
- branch=renovate/major-all-dependencies
- branch=renovate/major-all-devdependencies

**Reply by viceice** [reply]
Make sure to run with `LOG_LEVEL=debug` as that is a debug message

**Reply by rarkins** [reply]
I can't help you further until you can print your config. I just checked locally and it works fine with this command : `renovate --print-config=true org/repo`

The config that's printed is with this message: `DEBUG: Full resolved config including presets (repository=org/repo)`

**Reply by hieroshima** [reply]
```json
{
             "depTypeList": ["devDependencies"],
             "groupName": "devDependencies",
             "groupSlug": "all-devDependencies",
             "labels": ["devDependency"]
           },
           {
             "depTypeList": ["dependencies"],
             "groupName": "dependencies",
             "groupSlug": "all-dependencies",
             "labels": ["dependency"]
           },
           {"updateTypes": ["major"], "addLabels": ["major"]},
           {"updateTypes": ["minor"], "addLabels": ["minor"]},
           {"updateTypes": ["patch"], "addLabels": ["patch"]},
           {
             "packagePatterns": ["^angular"],
             "groupName": "angular",
             "groupSlug": "angular-update",
             "addLabels": ["angular"]
           },
           {
             "packagePatterns": ["^ngrx"],
             "groupName": "ngrx",
             "groupSlug": "ngrx-update",
             "addLabels": ["ngrx"]
           }
```

This is the last four groups from the printed config, so I think that this is ok when I assume that the last group is the most important? Maybe that groups related to updateTypes breaking something?

**Reply by hieroshima** [reply]
It works... the problem was in patternPackage... All angular packages start `@angular`, so I fixed packagePattern `^@angular` and now I have separate angular PR with devDep only and separate angular with dependencies. Is it possible to put both devDep and dep in one PR in case of angular?

Also, I would like to ask about the possibility of those scenarios together:

- All devDependency without splitting to major/minor/patch -- this is already done
- All major dependencies in separate PR for each major dep
- All minor/patch dependencies in one PR for all minor/patch dependency
- All angular devDep and Dep together in one PR without splitting to major/minor/patch

I would like have this scenarios in one repository and I wonder if it is possible, because it is related with the 

"separateMultipleMajor": false,
  "separateMinorPatch": false,

settings which I think can't be used inside packageRules?

**Reply by rarkins** [reply]
Those settings can be used inside packageRules but not if you combine with `updateTypes` - use two rules instead of one.

**Reply by hieroshima** [reply]
Ok, so how I can have everything related to devDependancy in one PR? If I have that packageRules:

```json
"separateMultipleMajor": false,
  "separateMinorPatch": false,
  "packageRules": [
    {
      "depTypeList": ["devDependencies"],
      "updateTypes": ["major", "minor", "patch"],
      "groupName": "devDependencies",
      "groupSlug": "all-devDependencies",
      "labels": ["devDependency"]
    },
```
It still create two PR/branches - `all-devDependancies` and `major-all-devDependancies`. Is there a way to have everything in one PR?

**Reply by viceice** [reply]
https://docs.renovatebot.com/configuration-options/#separatemajorminor

**Reply by hieroshima** [reply]
Perfect, thank you @rarkins and @viceice for help
